adapters/tpmn/tpmn.go

```diff
@@ -0,0 +1,153 @@
+package tpmn
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/prebid/openrtb/v19/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+// TpmnAdapter struct
+type adapter struct {
+	uri string
+}
+
+// MakeRequests makes the HTTP requests which should be made to fetch bids from TpmnBidder.
+func (rcv *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	validImps, errs := getValidImpressions(request, reqInfo)
+	if len(validImps) == 0 {
+		return nil, errs
+	}
+
+	request.Imp = validImps
+
+	requestBodyJSON, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	return []*adapters.RequestData{{
+		Method:  http.MethodPost,
+		Uri:     rcv.uri,
+		Body:    requestBodyJSON,
+		Headers: headers,
+	}}, errs
+}
+
+// getValidImpressions validate imps and check for bid floor currency. Convert to EUR if necessary
+func getValidImpressions(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]openrtb2.Imp, []error) {
+	var errs []error
+	var validImps []openrtb2.Imp
+
+	for _, imp := range request.Imp {
+		if err := preprocessBidFloorCurrency(&imp, reqInfo); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if err := preprocessExtensions(&imp); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		validImps = append(validImps, imp)
+	}
+	return validImps, errs
+}
+
+func preprocessExtensions(imp *openrtb2.Imp) error {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var tpmnBidderExt openrtb_ext.ExtImpTpmn
+	if err := json.Unmarshal(bidderExt.Bidder, &tpmnBidderExt); err != nil {
+		return &errortypes.BadInput{
+			Message: "Wrong TPMN bidder ext: " + err.Error(),
+		}
+	}
```

Comment:
I do not think you need this check here. This check is not doing anything but unmarshalling json to `openrtb_ext.ExtImpTpmn` type that is unnecessary check. Requesting you to please remove it.

Reply:
update commit :
[f861d08](https://github.com/prebid/prebid-server/pull/2947/commits/f861d080a269655a4ac44d57daa2a10b7bb0256d)